dataavailability/datacommittee/datacommittee.go

```diff
@@ -0,0 +1,292 @@
+package datacommittee
+
+import (
+	"crypto/ecdsa"
+	"errors"
+	"fmt"
+	"math/big"
+	"math/rand"
+	"sort"
+	"strings"
+
+	"github.com/0xPolygon/cdk-data-availability/client"
+	daTypes "github.com/0xPolygon/cdk-data-availability/types"
+	"github.com/0xPolygonHermez/zkevm-node/etherman/smartcontracts/polygondatacommittee"
+	"github.com/0xPolygonHermez/zkevm-node/log"
+	"github.com/ethereum/go-ethereum/accounts/abi/bind"
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/crypto"
+	"github.com/ethereum/go-ethereum/ethclient"
+	"golang.org/x/net/context"
+)
+
+// DataCommitteeMember represents a member of the Data Committee
+type DataCommitteeMember struct {
+	Addr common.Address
+	URL  string
+}
+
+// DataCommittee represents a specific committee
+type DataCommittee struct {
+	AddressesHash      common.Hash
+	Members            []DataCommitteeMember
+	RequiredSignatures uint64
+}
+
+// DataCommitteeBackend implements the DAC integration
+type DataCommitteeBackend struct {
+	dataCommitteeContract      *polygondatacommittee.Polygondatacommittee
+	l2Coinbase                 common.Address
+	privKey                    *ecdsa.PrivateKey
+	dataCommitteeClientFactory client.IClientFactory
+
+	committeeMembers        []DataCommitteeMember
+	selectedCommitteeMember int
+	ctx                     context.Context
+}
+
+// New creates an instance of DataCommitteeBackend
+func New(
+	l1RPCURL string,
+	dataCommitteeAddr common.Address,
+	l2Coinbase common.Address,
+	privKey *ecdsa.PrivateKey,
+	dataCommitteeClientFactory client.IClientFactory,
+) (*DataCommitteeBackend, error) {
+	ethClient, err := ethclient.Dial(l1RPCURL)
+	if err != nil {
+		log.Errorf("error connecting to %s: %+v", l1RPCURL, err)
+		return nil, err
+	}
+	dataCommittee, err := polygondatacommittee.NewPolygondatacommittee(dataCommitteeAddr, ethClient)
+	if err != nil {
+		return nil, err
+	}
+	return &DataCommitteeBackend{
+		dataCommitteeContract:      dataCommittee,
+		l2Coinbase:                 l2Coinbase,
+		privKey:                    privKey,
+		dataCommitteeClientFactory: dataCommitteeClientFactory,
+		ctx:                        context.Background(),
+	}, nil
+}
+
+// Init loads the DAC to be cached when needed
+func (d *DataCommitteeBackend) Init() error {
+	committee, err := d.getCurrentDataCommittee()
+	if err != nil {
+		return err
+	}
+	selectedCommitteeMember := -1
+	if committee != nil {
+		d.committeeMembers = committee.Members
+		if len(committee.Members) > 0 {
+			selectedCommitteeMember = rand.Intn(len(committee.Members)) //nolint:gosec
+		}
+	}
+	d.selectedCommitteeMember = selectedCommitteeMember
+	return nil
+}
+
+const unexpectedHashTemplate = "missmatch on transaction data for batch num %d. Expected hash %s, actual hash: %s"
+
+// GetData returns the data from the DAC. It checks that it matches with the expected hash
+func (d *DataCommitteeBackend) GetData(batchNum uint64, hash common.Hash) ([]byte, error) {
```

Comment:
By looking at this `for` loop, seems there can be cases where it can run forever. Should we add some timeout to stop its execution after some time?

Reply:
the code is hard to follow, but should always break after reaching the initial committee member that has been queried